src/zon.zig

```diff
@@ -0,0 +1,831 @@
+    }
+
+    const field_names = struct_info.field_names.get(ip);
+    for (field_values, field_names) |*value, name| {
+        if (value.* == .none) return self.fail(
+            .{ .node_abs = node.getAstNode(self.file.zoir.?) },
+            "missing field {}",
+            .{name.fmt(ip)},
+        );
+    }
+
+    return self.sema.pt.intern(.{ .aggregate = .{ .ty = res_ty.toIntern(), .storage = .{
+        .elems = field_values,
+    } } });
+}
+
+fn lowerPointer(self: LowerZon, node: Zoir.Node.Index, res_ty: Type) !InternPool.Index {
+    const ip = &self.sema.pt.zcu.intern_pool;
+    const gpa = self.sema.gpa;
+
+    const ptr_info = res_ty.ptrInfo(self.sema.pt.zcu);
+
+    if (ptr_info.flags.size != .Slice) {
+        return self.fail(
+            .{ .node_abs = node.getAstNode(self.file.zoir.?) },
+            "non slice pointers are not available in ZON",
+            .{},
+        );
+    }
+
+    // String literals
+    const string_alignment = ptr_info.flags.alignment == .none or ptr_info.flags.alignment == .@"1";
+    const string_sentinel = ptr_info.sentinel == .none or ptr_info.sentinel == .zero_u8;
+    if (string_alignment and ptr_info.child == .u8_type and string_sentinel) {
+        switch (node.get(self.file.zoir.?)) {
+            .string_literal => |val| {
+                const ip_str = try ip.getOrPutString(gpa, self.sema.pt.tid, val, .maybe_embedded_nulls);
+                const str_ref = try self.sema.addStrLit(ip_str, val.len);
+                return (try self.sema.coerce(
+                    self.block,
+                    res_ty,
+                    str_ref,
+                    try self.lazySrcLoc(.{ .node_abs = node.getAstNode(self.file.zoir.?) }),
+                )).toInterned().?;
+            },
+            else => {},
+        }
+    }
+
+    // Slice literals
+    const elem_nodes: Zoir.Node.Index.Range = switch (node.get(self.file.zoir.?)) {
+        .array_literal => |nodes| nodes,
+        .empty_literal => .{ .start = node, .len = 0 },
+        else => return self.fail(
+            .{ .node_abs = node.getAstNode(self.file.zoir.?) },
+            "expected type '{}'",
+            .{res_ty.fmt(self.sema.pt)},
+        ),
+    };
+
+    const elems = try self.sema.arena.alloc(InternPool.Index, elem_nodes.len + @intFromBool(ptr_info.sentinel != .none));
+
+    for (0..elem_nodes.len) |i| {
+        elems[i] = try self.lowerExpr(elem_nodes.at(@intCast(i)), Type.fromInterned(ptr_info.child));
+    }
+
+    if (ptr_info.sentinel != .none) {
+        elems[elems.len - 1] = ptr_info.sentinel;
+    }
+
+    const array_ty = try self.sema.pt.intern(.{ .array_type = .{
+        .len = elems.len,
+        .sentinel = ptr_info.sentinel,
+        .child = ptr_info.child,
+    } });
+
+    const array = try self.sema.pt.intern(.{ .aggregate = .{
+        .ty = array_ty,
+        .storage = .{ .elems = elems },
+    } });
+
+    const many_item_ptr_type = try ip.get(gpa, self.sema.pt.tid, .{ .ptr_type = .{
```

Comment:
Huh, any reason these lines use `ip.get` while the above use `pt.intern`?

Reply:
Ah nope this was a mistake, good catch!